service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+		r.RLock()
+		requestTableVersion := r.tableVersion
+		r.RUnlock()
+
+		shouldWait := currentPageToken == nil && requestTableVersion != 0
+
+		resp, err := r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+			LastKnownTableVersion: requestTableVersion,
+			NextPageToken:         currentPageToken,
+			PageSize:              loadServicesPageSize,
+			Wait:                  shouldWait,
+		})
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// indicates table was updated during paging, so reset and start from the beginning
+				currentPageToken = nil
+				pagingWasReset = true
+				r.invalidateCachedServicesLocked()
+				continue
+			}
+			// for all other errors, fall back to reading directly from persistence
+			if currentPageToken == nil && !pagingWasReset {
+				r.Lock()
+			}
+			persistenceErr := r.loadServicesLocked(ctx)
+			if persistenceErr == nil {
+				r.setServiceDataReady(persistenceErr)
+			}
+			r.Unlock()
+			return persistenceErr
+		}
+
+		if shouldWait && requestTableVersion == resp.TableVersion {
+			// long poll returned with no changes
+			return nil
+		}
+
+		if currentPageToken == nil && !pagingWasReset {
+			// acquire lock once upon getting first page and release once all pagination is complete
+			r.Lock()
+		}
+
+		if shouldWait {
+			// long poll returned new data, starting with first page
+			r.invalidateCachedServicesLocked()
+		}
+
+		currentPageToken = resp.NextPageToken
+		r.tableVersion = resp.TableVersion
+		r.updateCachedServicesLocked(resp.Entries)
+
+		if resp.NextPageToken == nil {
+			break
+		}
+	}
+
+	r.setServiceDataReady(ctx.Err())
+	r.Unlock()
+	return ctx.Err()
+}
+
+// Loads all incoming services into memory from persistence.
+// Used as a fallback if matching service is unavailable.
+// Always replaces entire in-memory view by paging from the start.
+func (r *registry) loadServicesLocked(ctx context.Context) error {
+	r.invalidateCachedServicesLocked()
+
+	var currentPageToken []byte
+
+	for ctx.Err() == nil {
+		resp, err := r.persistence.ListNexusIncomingServices(ctx, &p.ListNexusIncomingServicesRequest{
+			LastKnownTableVersion: r.tableVersion,
+			NextPageToken:         currentPageToken,
+			PageSize:              loadServicesPageSize,
+		})
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// indicates table was updated during paging, so reset and start from the beginning
+				currentPageToken = nil
+				r.invalidateCachedServicesLocked()
+				continue
+			}
+			return err
+		}
+
+		currentPageToken = resp.NextPageToken
+		r.tableVersion = resp.TableVersion
+		r.updateCachedServicesLocked(resp.Entries)
+
+		if resp.NextPageToken == nil {
+			break
+		}
+	}
+
+	return ctx.Err()
+}
+
+func (r *registry) refreshServices(ctx context.Context) error {
+	refreshFn := func(ctx context.Context) error {
+		_, err := r.serviceDataReady.Get(ctx)
+		if err != nil {
+			return err
+		}
```

Comment:
You can move this inline to before the loop in `refreshServices` (line 342). The condition isn't going to be invalidated.